Refactors

Anton-Latukha opened the pull request:
Reducing some current YAGNIs.

Starting reducing some superfluous things in the code (https://github.com/haskell-nix/hnix/issues/879)

* `Result` type is really an euphemism to `Either`. `Left` `Right` is on the instinctive understanding for Hakellers, and using it enables further readability and plasticity to the code. 
Currently made `Result`, what it is - an euphemism to `Either`. Probably we may even just remove the custom type without any detriment, it probably would just make code even easier to read, since the `Either` semantics already fully embody what `Result` is, and the habitat of the function that brings `Either` shows where it came from. Outside of the module that exposes the data type `Result` is literally does not appear not once in type signatures anywhere over the project.

* `forceEff` changed a little, no reason to lock and then do preparation actions that do not required locking the whole process. So now doing preparing first and then locking right when it is needed.

* Please note that `force{,Eff}` implementations internally are now the same.

* `queryM` became non-locking, since read from static structures in a static language does not require locking operations.

---

Update:

* notice the implementation transformation for `further` - now it really shows the nature that all there was - it really simply defers the thunk. The question here is: since locking was not implemented for it, does `atomicModifyVar` needs thunk locking, or its atomicity allows this. But anyway, `further` family of functions used just 1-nce in a `inform` type cass instance that is not used at all.

* we arrived at that `force{,Eff}` are exactly 1 function, they are united currently under `forceMain`, the future (a cascade other MonadThunk instances above, for example, and future in general) would show does `class MonadThunk` needs `forceEff`, or `forceEff` is just `force`.

Anton-Latukha committed:
Repl: refactor

Anton-Latukha committed:
Exec: refactor; add ApplicativeDo

Anton-Latukha committed:
Utils: rename fun (`if*` -> `when*`) (improves readability)

For the `if` FP programmers always internally psychologically ask `& else?`.
And since this functions implicitly return as an else a stub instead of
processing, for what `when*` is used in the language - fits better.

Anton-Latukha committed:
Main: layout

Anton-Latukha committed:
Parser: `Result` is just an euphemism to `Either`

Implementation show that we so far not needed the custom type for it. Right and
Left already embodies the whole semantic meaning, since Either is mostly used as
Right for computations being successful, and Left for failure. So why bother
everyone with remembering the custom constructors and what they are for if they
are not needed, Either is already know to everybody.

Current implementation by itself shows that `Result` custom type is being YAGNI,
since replacing it with Either changes nothing.
So it relates to the topic of https://github.com/haskell-nix/hnix/issues/879.

If we would need ad-hock abilities to "our custom class" - the
`TypeSynonymInstances` are there for us.

I did the main work abstracting implementation from the type and constructors
here, so if we would want in the future to change the type - we would just need
to replace the `either` function with `result` function and that is all.

But just using `either` is more consise.

Type synonym left the signature information there. I'd even considered to remove
`Result` all togather, since the more I talk about it the more it becomes clear
that its use is just what Haskellers use `Either` for.

Anton-Latukha committed:
Utils: add `both`

Anton-Latukha committed:
{Pretty, Reduce, PrettyParseTests}: m refactors

Anton-Latukha committed:
Reduce: m refactor

Anton-Latukha committed:
Thunk.Basic: forceEff: refactor; enabling `ApplicativeDo`

Notice, here in `forceEff` I moved the thunk locking closer to where it is
needed, so function does a couple of preparation computations before locking the
thunk to operate on it.

Also notice that the `force` add `forceEff` look like 1 function, and as
`forceEff` also probably needs to check for exception when doing an action -
they are literally the same function.

Anton-Latukha committed:
Thunk.Basic: queryM: make it non-blocking

So far `queryM` is not used. And locking the thunk computation on static-copying
structures to check if variable is computed is weird. Lets allow reading from
static structure until we would need locking during reads.

Anton-Latukha committed:
Thunk.Basic: forceEff: update exception processing

Seems logical that `forceEff` needs to have the exception catcher on `action`
and "Loop detected" is just a ThunkLoop.

So now `forceEff == force`

Ericson2314 commented on src/Nix/Options/Parser.hs:
```diff
@@ -62,7 +62,7 @@ nixOptions current =
     <*> switch
         (  long "values"
-        <> help "Enable reporting of value provenance in error messages"
+        <> help "Enable reporting of value provenance in fail messages"
```
unintentional replace?

Anton-Latukha replied:
Thank you. This one is indeed.

Anton-Latukha committed:
Thunk.Basic: Utils: add internall `deferred` fun

Anton-Latukha committed:
Thunk.Basic: utils: deferred: upd type; use fun for `further`

Now it shows that `further` body is actually a `const`.

Anton-Latukha committed:
Utils: add `dup`; Thunk.Basic: further: define with a `dup`

We saw that `further` does not even need to strictly pattern match on a thunk,
since the results of bothe matches would be the same, so we can skip matching.

Anton-Latukha committed:
Thunk.Basic: force{,Eff}: unify implementation as `forceMain`

Anton-Latukha committed:
Options.Parser: refactor

Just organizing and layouting

Anton-Latukha committed:
cabal: flag optimize: add ApplicativeDo

At this point it optimizes things greatly.

Anton-Latukha committed:
cabal: flag optimize: add O3 as it works-out good

Anton-Latukha committed:
cabal: enable optimizations by default

Anton-Latukha committed:
cabal: layout

Anton-Latukha committed:
cabal: hnix: disable default -rtsopts

Anton-Latukha committed:
Utils: loebM: function is used much, optimization

It can be inlined. It also appears just 2 times in the code, but in Eval code.

Anton-Latukha committed:
cabal: stopping GHC 8.4 support

https://github.com/haskell-nix/hnix/issues/885

Anton-Latukha committed:
Value: fx MonadFail for GHC 8.6

Anton-Latukha committed:
Thunk.Basic: refactoring the thunk lock/unlock procedure

Anton-Latukha committed:
Thunk.Basic: more refactor for locking mechanism; module organization

Anton-Latukha committed:
Reduce: refactor

Anton-Latukha committed:
treewide: lazify handling: in place replacable (error -> fail)

This is was was possible to replace right away in-place.

Also added `MonadFail` requirement to `MonadDataErrorContext t f m`, as it was
obvious addition to allow to demote some `errors` into `fail`.

Reference:
https://www.haskellforall.com/2019/12/prefer-to-use-fail-for-io-exceptions.html

Anton-Latukha committed:
ChangeLog: add patch updates

Anton-Latukha committed:
treewide: clean-up `ApplicativeDo` after it enabled by default
